Use RenderAttachment for wide-layout attachments, with some modified layout and rendering

squelart opened the pull request:
#### 333b4b2030a128aebb08c23403fb50e4465172ca
<pre>
Use RenderAttachment for wide-layout attachments, with some modified layout and rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=256637">https://bugs.webkit.org/show_bug.cgi?id=256637</a>
rdar://108157331

Reviewed by Alan Baradlay.

A lot of the editor code interacting with attachments relies on the renderer being exactly `RenderAttachment`, which led to some misbehavior when the wide-layout attachment was using the more generic `RenderBlockFlow`.
So now the renderer is always `RenderAttachment`, but the layout and rendering paths redirect to special wide-layout-only functions.

* LayoutTests/fast/attachment/cocoa/wide-attachment-save-event-expected.txt:
* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::createElementRenderer):
Always return a `RenderAttachment`, like it was before introducing the wide-layout styling.

(WebCore::HTMLAttachmentElement::insertedIntoAncestor):
Because of how the root element does the layout, the shadow root element&apos;s margins were ignored, so now the margins are added here in the root element.

* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
Since the margin is now in the top-level element outside the shadow tree, it is not needed here anymore.

(div#attachment-container::selection): Deleted.
The selection coloring is handled by the `RenderReplaced` painting code, so this styling was ignored anyway.

* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::RenderAttachment):
(WebCore::RenderAttachment::layoutWideLayoutAttachmentOnly):
(WebCore::RenderAttachment::layout):
Layout the wide-layout shadow tree if present, the &quot;replaced&quot; and shadow content layouts are still performed to handle all the necessary sizing and optional image controls.

(WebCore::RenderAttachment::paintWideLayoutAttachmentOnly const):
New wide-layout paint code, handling the full Foreground and Selection phases and painting everything in the shadow tree.

* Source/WebCore/rendering/RenderAttachment.h:
Cleaned up &quot;Shadow&quot; functions, to separate just &quot;shadow controls&quot; (like image controls) and any &quot;shadow content&quot; that may include the wide-layout shadow tree.
Also let the `RenderReplaced` painting code handle the selection tint.

* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::paintAttachment):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):
When actually painting a wide-layout attachment, skip the legacy painting.

Canonical link: <a href="https://commits.webkit.org/264914@main">https://commits.webkit.org/264914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0dd380f473287c94c6bdc3917320e4128f61199f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9012 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9518 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10665 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8974 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9020 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11286 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9267 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11826 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9158 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10134 "1 flakes 2 failures") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10824 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8239 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15719 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8538 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8387 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11708 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8883 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7244 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8127 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2209 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12342 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8633 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -372,6 +372,36 @@ void HTMLAttachmentElement::invalidateRendering()
+        return false;
+
+    if (auto* shadowRenderer = m_containerElement->renderer()) {
+        PaintInfo shadowPaintInfo = paintInfo;
```
This seems like a layering violation. I don't think DOM/HTML should know about painting.

squelart replied:
Alright, I'll move these to `RenderAttachment`, but I'll have to expose the container element to help.

pxlcoder commented on Source/WebCore/rendering/RenderAttachment.cpp:
```diff
@@ -43,10 +43,18 @@ namespace WebCore {
+#if (defined(WTF_PLATFORM_IOS_FAMILY) && WTF_PLATFORM_IOS_FAMILY)
+static constexpr IntSize attachmentWideLayoutIntrisicSize { 338, 92 };
+#else
+static constexpr IntSize attachmentWideLayoutIntrisicSize { 266, 80 };
```
Do we really need these, or can we pull from the `RenderStyle`?

squelart replied:
`RenderStyle` didn't have these values, so instead I'm using the size of the laid-out shadow tree.

nt1m reviewed (dismissed):
I don't think we should land this hack, that is reinventing the rendering code for shadow content (which means it can easily become out of date).

Good next steps would be to create web content (not relying on `attachment`, using shadow DOM) that reproduces this bug, and start from there. This is likely a bug that also affects web content.

squelart committed:
Use RenderAttachment for wide-layout attachments, with some modified layout and rendering
https://bugs.webkit.org/show_bug.cgi?id=256637
rdar://108157331

Reviewed by Alan Baradlay.

A lot of the editor code interacting with attachments relies on the renderer being exactly `RenderAttachment`, which led to some misbehavior when the wide-layout attachment was using the more generic `RenderBlockFlow`.
So now the renderer is always `RenderAttachment`, but the layout and rendering paths redirect to special wide-layout-only functions.

* LayoutTests/fast/attachment/cocoa/wide-attachment-save-event-expected.txt:
* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::createElementRenderer):
Always return a `RenderAttachment`, like it was before introducing the wide-layout styling.

(WebCore::HTMLAttachmentElement::insertedIntoAncestor):
Because of how the root element does the layout, the shadow root element's margins were ignored, so now the margins are added here in the root element.

* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-container):
Since the margin is now in the top-level element outside the shadow tree, it is not needed here anymore.

(div#attachment-container::selection): Deleted.
The selection coloring is handled by the `RenderReplaced` painting code, so this styling was ignored anyway.

* Source/WebCore/rendering/RenderAttachment.cpp:
(WebCore::RenderAttachment::RenderAttachment):
(WebCore::RenderAttachment::layoutWideLayoutAttachmentOnly):
(WebCore::RenderAttachment::layout):
Layout the wide-layout shadow tree if present, the "replaced" and shadow content layouts are still performed to handle all the necessary sizing and optional image controls.

(WebCore::RenderAttachment::paintWideLayoutAttachmentOnly const):
New wide-layout paint code, handling the full Foreground and Selection phases and painting everything in the shadow tree.

* Source/WebCore/rendering/RenderAttachment.h:
Cleaned up "Shadow" functions, to separate just "shadow controls" (like image controls) and any "shadow content" that may include the wide-layout shadow tree.
Also let the `RenderReplaced` painting code handle the selection tint.

* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::paintAttachment):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):
When actually painting a wide-layout attachment, skip the legacy painting.

Canonical link: https://commits.webkit.org/264914@main
